File: LeetCode/76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int m = s.size(), n = t.size();
        int l = 0, r = 0, minlen = INT_MAX, numPos = n, minStart = -1;
        vector<int> window (58, 0);
        string sol = "";

        for (int i = 0; i < n; i++) {
            window[t.at(i) - 'A']++;
        } //for

        if (n > m) return sol;

        while (r < m) {
            window[s.at(r) - 'A']--;
            if (window[s.at(r) - 'A'] >= 0) numPos--;
            while (numPos == 0) {
                if (r - l < minlen) {
                    minlen = r - l;
                    minStart = l;
                    //sol = s.substr(l, r - l + 1);
                } //if
                window[s.at(l) - 'A']++;
                if (window[s.at(l) - 'A'] > 0) numPos++;
                l++;
            } //while
            r++;
        } //while

        return minlen == INT_MAX ? "" : s.substr(minStart, ++minlen);
    } //minWindow
};

File: LeetCode/76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) return "";
        
        int [] tfreq = new int [128];
        int [] window = new int [128];
        int l = 0;
        int lMin = Integer.MAX_VALUE, gMin = Integer.MAX_VALUE;
        int matches = 0;
        String sol = "";
        
        //populate tFreq
        for (int i = 0; i < t.length(); i++)
            tfreq[t.charAt(i)]++;
        
        //populate window
        for (int i = 0; i < t.length() - 1; i++) {
            window[s.charAt(i)]++;
            if (window[s.charAt(i)] <= tfreq[s.charAt(i)]) matches++;
        } //for
        
        //main
        for (int r = t.length() - 1; r < s.length(); r++) {
            window[s.charAt(r)]++;
            if (window[s.charAt(r)] <= tfreq[s.charAt(r)]) matches++;
            
            while (matches == t.length()) {
                lMin = r - l + 1;
                if (lMin < gMin) {
                    gMin = lMin;
                    sol = s.substring(l,r+1);
                } //if
                window[s.charAt(l)]--;
                if (window[s.charAt(l)] < tfreq[s.charAt(l)]) matches--;
                l++;
            } //while
        } //for
        
        System.gc();
        return sol;
    } //minWin
} //Sol